lib/Sema/BuilderTransform.cpp

```diff
@@ -741,11 +741,124 @@ class BuilderClosureVisitor
     return visit(caseStmt->getBody());
   }
 
+  VarDecl *visitForEachStmt(ForEachStmt *forEachStmt) {
+    // for...in statements are handled via buildArray(_:); bail out if the
+    // builder does not support it.
+    if (!builderSupports(ctx.Id_buildArray)) {
+      if (!unhandledNode)
+        unhandledNode = forEachStmt;
+      return nullptr;
+    }
+
+    // For-each statements require the Sequence protocol. If we don't have
+    // it (which generally means the standard library isn't loaded), fall
+    // out of the function-builder path entirely to let normal type checking
+    // take care of this.
+    auto sequenceProto = TypeChecker::getProtocol(
+        dc->getASTContext(), forEachStmt->getForLoc(),
+        KnownProtocolKind::Sequence);
+    if (!sequenceProto) {
+      if (!unhandledNode)
+        unhandledNode = forEachStmt;
+      return nullptr;
+    }
+
+    // Generate constraints for the loop header. This also wires up the
+    // types for the patterns.
+    auto target = SolutionApplicationTarget::forForEachStmt(
+        forEachStmt, sequenceProto, dc, /*bindPatternVarsOneWay=*/true);
+    if (cs) {
+      if (cs->generateConstraints(target, FreeTypeVariableBinding::Disallow)) {
+        hadError = true;
+        return nullptr;
+      }
+
+      cs->setSolutionApplicationTarget(forEachStmt, target);
+    }
+
+    // Visit the loop body itself.
+    VarDecl *bodyVar = visit(forEachStmt->getBody());
+    if (!bodyVar)
+      return nullptr;
+
+    // If there's no constraint system, there is nothing left to visit.
+    if (!cs)
+      return nullptr;
+
+    // Form a variable of array type that will capture the result of each
+    // iteration of the loop. We need a fresh type variable to remove the
+    // lvalue-ness of the array variable.
+    SourceLoc loc = forEachStmt->getForLoc();
+    VarDecl *arrayVar = buildVar(loc);
+    Type arrayElementType = cs->createTypeVariable(
+        cs->getConstraintLocator(forEachStmt), 0);
+    cs->addConstraint(
+        ConstraintKind::Equal, cs->getType(bodyVar), arrayElementType,
+        cs->getConstraintLocator(
+          forEachStmt, ConstraintLocator::RValueAdjustment));
+    Type arrayType = ArraySliceType::get(arrayElementType);
+    cs->setType(arrayVar, arrayType);
+
+    // Form an initialization of the array to an empty array literal.
+    Expr *arrayInitExpr = ArrayExpr::create(ctx, loc, { }, { }, loc);
+    cs->setContextualType(
+        arrayInitExpr, TypeLoc::withoutLoc(arrayType), CTP_CannotFail);
+    arrayInitExpr = cs->generateConstraints(arrayInitExpr, dc);
+    if (!arrayInitExpr) {
+      hadError = true;
+      return nullptr;
+    }
+    cs->addConstraint(
+        ConstraintKind::Equal, cs->getType(arrayInitExpr), arrayType,
+        cs->getConstraintLocator(arrayInitExpr));
```

Comment:
Shouldn't constraint have `array expr -> contextual type` locator in this case because it's exactly the same as a typed pattern e.g. `var _: [$T] = []`?

Reply:
Ah, sure!

Reply:
https://github.com/apple/swift/pull/31624